### PR TITLE
message: port advanced rpc compression from scylla-enterprise.git to scylladb.git

### PR DESCRIPTION
**Pull request by michoecho**
This is a forward port (from scylla-enterprise) of additional compression options (zstd, dictionaries shared across messages) for inter-node network traffic. It works as follows:

After the patch, messaging_service (Scylla's interface for all inter-node communication)
compresses its network traffic with compressors managed by
the new advanced_rpc_compression::tracker. Those compressors compress with lz4,
but can also be configured to use zstd as long as a CPU usage limit isn't crossed.

A precomputed compression dictionary can be fed to the tracker. Each connection
handled by the tracker will then start a negotiation with the other end to switch
to this dictionary, and when it succeeds, the connection will start being compressed using that dictionary.

All traffic going through the tracker is passed as a single merged "stream" through dict_sampler.
dictionary_service has access to the dict_sampler.
On chosen nodes (in the "usual" configuration: the Raft leader), it uses the sampler to maintain
a random multi-megabyte sample of the sampler's stream. Every several minutes,
it copies the sample, trains a compression dictionary on it (by calling zstd's
training library via the alien_worker thread) and publishes the new dictionary
to system.dicts via Raft's write_mutation command.

This update triggers (eventually) a callback on all nodes, which feeds the new dictionary
to advanced_rpc_compression::tracker, and this switches (eventually) all inter-node connections
to this dictionary.

### CODE REVIEW

**Comment by tchaikov** [comment]
```
/home/runner/work/scylladb/scylladb/docs/operating-scylla/procedures/config-change/advanced-internode-compression.rst: WARNING: document isn't included in any toctree
```

probably we failed to port some changes from 76f1ee6785ad997c4f3c4d58cfd827c712e0df1a

File: conf/scylla.yaml [hunk 1]
```diff
@@ -477,6 +477,47 @@ commitlog_total_space_in_mb: -1
...
+# by setting `when_leader` on all nodes, or specify one manually by setting `always`
+# on one node and `never` on others.
+# 
+# rpc_dict_training_when: never
+
+# A number in range [0.0, 1.0] specifying the share of CPU which can be spent
+# by this node on compressing inter-node traffic with zstd.
+#
+# Depending on the workload, enabling zstd might have a drastic negative
+# effect on performace, so it shouldn't be done lightly.
```
Comment (tchaikov) [inline conf/scylla.yaml:517]: could take this opportunity to fix the typo.

File: docs/dev/advanced_rpc_compression.md [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+During dictionary negotiation, nodes identify dictionaries by a `<timestamp, origin, content_sha256>` triple.
+
+If two different dictionaries with a different content but the same SHA256 are ever
+published with the same `<timestamp, origin>`, the negotiation might result in a corrupted
+connection. No countermeasures against this have been implemented as of now.
+
+### Negotiation
+
+Dictionary compression requires that the sender and the receiver both use the same dictionary.
+In particular, the sender can only start using a new dictionary after the receiver aquires
```
Comment (tchaikov) [inline docs/dev/advanced_rpc_compression.md:146]: ditto. might want to fix the typo here.

File: docs/dev/advanced_rpc_compression.md [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+send the new dictionary over the connection whenever it wants to switch. This requires no
+inter-node coordination. However, it's relatively expensive -- the number of connections is quadratic
+in the size of the cluster, and this method requires each connection to transfer the entire
+dictionary. Ideally, each node should only have to download each new dictionary only once.
+
+#### Description
+
+The negotiation algorithm we currently use works like this (it is run for each connection direction independently):
+
+There are two peers: sender (compresses) and receiever (decompresses). They communicate via an ordered
```
Comment (tchaikov) [inline docs/dev/advanced_rpc_compression.md:162]: ditto.

**Comment by michoecho** [comment]
v2: Addressed Kefu's comments.
- Fixed 3 typos.
- Added `docs/operating-scylla/procedures/config-change/advanced-internode-compression.rst` to `docs/operating-scylla/procedures/config-change/index.rst`, same as in scylladb/scylla-enterprise@76f1ee6785ad997c4f3c4d58cfd827c712e0df1a

**Comment by michoecho** [comment]
@avikivity Ping for review/merge.

**Comment by avikivity** [comment]
Ugh the typo fixes are unfortunate, they will complicate the master->enterprise merge for me.

**Comment by michoecho** [comment]
> Ugh the typo fixes are unfortunate, they will complicate the master->enterprise merge for me.

Shall I revert them?

**Comment by avikivity** [comment]
> > Ugh the typo fixes are unfortunate, they will complicate the master->enterprise merge for me.
> 
> Shall I revert them?

No, I managed to resolve the problems.

I saw that scylladb/scylla-enterprise@e0800d1421031bc6540de6dfa50c6986678b11d4 missed the boat and has to be forward-ported separately.

**Pull request merged**